core/services/headtracker/heads.go

```diff
@@ -0,0 +1,113 @@
+package headtracker
+
+import (
+	"sort"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/smartcontractkit/chainlink/core/services/eth"
+)
+
+// Heads is a collection of heads. All methods are thread-safe.
+type Heads interface {
+	// LatestHead returns the block header with the highest number that has been seen, or nil.
+	LatestHead() *eth.Head
+	// HeadByHash returns a head for the specified hash, or nil.
+	HeadByHash(hash common.Hash) *eth.Head
+	// AddHeads adds newHeads to the collection, eliminates duplicates,
+	// sorts by head number, fixes parents and cuts off old heads (historyDepth).
+	AddHeads(historyDepth uint, newHeads ...*eth.Head)
+	// Count returns number of heads in the collection.
+	Count() int
+}
+
+type heads struct {
+	heads []*eth.Head
+	mu    sync.RWMutex
+}
+
+func NewHeads() Heads {
+	return &heads{}
+}
+
+func (h *heads) LatestHead() *eth.Head {
+	h.mu.RLock()
+	defer h.mu.RUnlock()
+
+	if len(h.heads) == 0 {
+		return nil
+	}
+	return h.heads[0]
+}
+
+func (h *heads) HeadByHash(hash common.Hash) *eth.Head {
+	h.mu.RLock()
+	defer h.mu.RUnlock()
+
+	for _, head := range h.heads {
+		if head.Hash == hash {
+			return head
+		}
+	}
+	return nil
+}
+
+func (h *heads) Count() int {
+	h.mu.RLock()
+	defer h.mu.RUnlock()
+
+	return len(h.heads)
+}
+
+func (h *heads) AddHeads(historyDepth uint, newHeads ...*eth.Head) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	headsMap := make(map[common.Hash]*eth.Head, len(h.heads)+len(newHeads))
+	for _, head := range append(h.heads, newHeads...) {
+		if head.Hash == head.ParentHash {
+			// shouldn't happen but it is untrusted input
+			continue
+		}
+		// copy all head objects to avoid races when a previous head chain is used
+		// elsewhere (since we mutate Parent here)
+		headCopy := *head
+		headCopy.Parent = nil // always build it from scratch in case it points to a head too old to be included
+		// map eliminates duplicates
+		headsMap[head.Hash] = &headCopy
+	}
+
+	heads := make([]*eth.Head, len(headsMap))
+	// unsorted unique heads
+	{
+		var i int
+		for _, head := range headsMap {
+			heads[i] = head
+			i++
+		}
+	}
+
+	// sort the heads
+	sort.SliceStable(heads, func(i, j int) bool {
+		// sorting from the highest number to lowest
+		return heads[i].Number > heads[j].Number
+	})
+
+	// cut off the oldest
+	if uint(len(heads)) > historyDepth {
+		heads = heads[:historyDepth]
+	}
+
+	// assign parents
+	for i := 0; i < len(heads)-1; i++ {
```

Comment:
This looks like a straight copy/paste of the old logic, which is good.

The `-1` is definitely important for reasons I can't quite recall, but I think it has to do with the fact that the final head in the chain must have no parent otherwise we end up with an infinitely long chain that never gets garbage collected.